AY2021S1-CS2103-T16-3/tp
# Functionality Increment: Add Command History

### What this does
This PR adds the command history functionality to ResiReg, allowing users to view their past commands in reverse chronological order. Users can also press <kbd>up</kbd> and <kbd>down</kbd> to cycle through the commands in the `CommandBox` (advance forward and backwards respectively).


Part of #30 

### How to test

1. Launch the application.
2. Type `history` in the command box and press <kbd>Enter</kbd>
3. Note that past commands appear in the results display
4. Press <kbd>up</kbd> and <kbd>down</kbd> to cycle through past commands entered

### Notes


An `InvalidationListenerList` class, along with methods to add + remove listeners and indicate modifications made to `ResiReg` are created to improve performance, by only saving whenever modifications pertaining to `ResiReg` or `UserPrefs` are made. 

This also ensures that the test `execute_storageThrowsIoException_throwsCommandException()` passes by allowing the history to be updated as the method call to update history is not interrupted due to the throwing and catching of the IoException thrown.

This change also means that `ReadOnlyResiReg` is now made `Observable` so that `ResiReg` can notify listeners on changes made to itself. 

I am planning to shift away from showing the commands in the results display to actually having a list layout for past commands. Perhaps this can be done after the UI changes.
